### DISCUSSION TITLE
Storing the meta data in DB

### DISCUSSION THREAD

**Discussion by atapas** [reactplay/react-play#16] [Ideas]
Currently, we are storing the meta data in a JSON file. How about persisting it into a DB?

Which DB? Let's discuss.

**Comment by atapas** [comment]
What are the DB Choices? Here are a few to list down,

- HarperDB
- PlanetScale
- Mongo Atlas
- Fauna
- Anything else?

**Reply by hasnainmakada-99** [reply]
Tapas Sir, I would like to add Firebase

**Comment by vasantisuthar** [comment]
Which schema structure should be implemented NoSQL or SQL?

**Reply by atapas** [reply]
Good Question. If we see from the requirements, we need, 

- Store the play meta data
- Registered users
- Likes for plays
- Comments and it can be hierarchical.

I can thinks of these from the storage requirements. What do you think would be suitable? I think NoSQL.

**Reply by vasantisuthar** [reply]
Yes, NoSQL would be a better option. Also, like the comments and likes for play should be in real-time, so can we consider using RethinkDB, it is a database that provides real-time updates to the application.

**Reply by atapas** [reply]
I love RethinkDB. It is a great real-time database and used extensively in my past projects.

I wouldn't consider RethinkDB for the following reasons. Let me know if you agree.

- There is NO cloud offering of RethinkDB. We need to host and maintain it on Heroku or AWS like PAAS.
- The development of RethinkDB is slow over the last 4-5 years. Also, there was a notice: https://rethinkdb.com/blog/rethinkdb-shutdown/
- We do not have a great need for Real-Time use cases.

What say?

**Reply by vasantisuthar** [reply]
Surely the reasons are true, we cannot consider RethinkDB now. 
And I didn't know much about HarperDB or PlanetScale but how about Firebase?
It provides a NoSQL cloud storage service

What do you think?

**Reply by atapas** [reply]
Surely worth looking.

**Comment by Abhishek-90** [comment]
I will be in favor of using Mongo Atlas or NoSQL database, for 2 reason, correct me if I am wrong.

1. For maintaining likes for Plays, using NoSQL schema will help us store details of users who have liked a particular play 
2. JSON format storing of Data in MongoDB will also help us to maintain hierarchy

**Reply by atapas** [reply]
@Abhishek-90 Agree. Both satisfy the criteria we are looking for. What are your thoughts about the plan and price? Do you think, we will scale there too if we go for it?

**Comment by hasnainmakada-99** [comment]
I am going to put a table ASAP, which will show us the pros and cons of every database we discussed over here, so will be good for us to ultimately think which DB we should consider to integrate

**Reply by atapas** [reply]
Thanks a lot. Have you used it before?

**Reply by hasnainmakada-99** [reply]
I am a little bit familiar with firebase, but have a general understanding of what the role is of each and every database 😃

**Reply by atapas** [reply]
> I am going to put a table ASAP, which will show us the pros and cons of every database we discussed over here, so will be good for us to ultimately think which DB we should consider to integrate

It will help us a lot. Thanks a bunch. Please keep cost as a factor as well. We are looking for opensource only option that will scale with our usages. Looking forward to the comparison @hasnainmakada-99

**Reply by hasnainmakada-99** [reply]
Will also keep the cost factor column in the table, will deliver it soon tapas sir

**Reply by atapas** [reply]
Btw, you can create the page here itself: https://github.com/atapas/react-play/wiki

**Reply by hasnainmakada-99** [reply]
Ok sir, will create it

**Reply by hasnainmakada-99** [reply]
Hello tapas sir,  Sir should I display the table here in the discussions only ?

**Reply by atapas** [reply]
Sure, whatever is comfortable to you.

**Reply by hasnainmakada-99** [reply]
ok sir

**Reply by hasnainmakada-99** [reply]
Tapas sir I've made tables which include 5 databases, I am gonna share it with you now

**Reply by hasnainmakada-99** [reply]
# ReactPlay.io

# List of Databases

| Harper DB |
| --- |
| Planet Scale |
| Mongo Atlas |
| Fauna |
| Firebase |

# Pros & Cons of every database

## Harper DB

| Pros | Cons | Cost | Harper DB Cloud |
| --- | --- | --- | --- |
| Easy to get started | Difficult to integrate with flask applications (As we are building on react it doesn’t matter to use) | The Cost of Harper DB Cloud... | 100% Customizable & Easy Deployment |
| Good Experience as per testing from the free trial | If coming from a SQL background, will be difficult to adapt Quickly |  | 1GB Free, After then 10GB at 2.5$/mo (190rs) |
| Customer/Tech Support is good compare to other companies | Less storage only 1gb is provided and only one instance allowed  |  | Then 250GB at 62.50$/mo (4750rs) |
|  |  |  | Then 1TB at 250$/mo (19000rs) |

## Planet Scale

| Pros | Cons | Cost |  |  |
| --- | --- | --- | --- | --- |
| Fully managed cloud native database built using Vitess and kubernetes, can access from anywhere using various cloud providers | Didn’t found out any flaws with Planet Scale | Planet Scale Developer (Free), limit one database per organization | Planet Scale Scalar | Planet Scale Enterprise |
| used by many big tech companies and small scale startups |  | 10 Gb Storage/Mo | 29$/mo (2200rs), perdatabase + usage cost | Custom pricing (to contact sales) |
| Has many alternatives such as amazon aurora, amazon RDS, google cloud SQL |  | 10 million row writes/mo | 25Gb storage & 100 bilion row reads/mo | unlimited database branches |
| Takes daily backup of databases |  | Daily backups & Community support | Auto Scaling | premium support |

## Mongo Atlas

| Pros | Cons | Cost |  |
| --- | --- | --- | --- |
| Performance level is good compare to others, as it allows specialized reads where we can map our data | Does have responsive management, we can not open the database from mobile & doesn’t works well with search parameters | MongoDB (Serverless) | MongoDB (Dedicated) |
| Possible to access JSON from anywhere without any language specialization | If entries go more than 150, we have to use the mongo shell platform & visualize the data | For serverless applications with low and variable traffic | Built for production applications with sophisticated workload requirements |
| Provides accurate information to users & takes backup of databases time to time | when working with clusters, it is not possible to download a direct copy of the information | Upto 1Tb of Storage |  |
|  |  | Starts from 0.30$/month (23rs) with millions reads | Starts from 57$/month (380rs), comes with 2 gb of ram  |

## Fauna DB

| Pros | Cons | Cost |  |
| --- | --- | --- | --- |
| Supports Both FQL (Fauna Query Language) & GraphQL | To write security rules we have to compulsory use GraphQL | Fauna DB Team | Fauna DB Business |
| FQL can be used to create databases, indexes, tables and many more in fauna | Consumed quite a time to get addapted to its domain specific FQL | 135$/month (10,300rs) | 450$/month (34,200rs) |
| Can be used as an external DB which we dont want to integrate in our app/website | Doesn’t allow to export the queries in JSON or CSV | Tech Support 24 business hours response time | same features as of team plan |
| Fast & Friendly Database |  | Unlimited Capacity & Multi factor authentication | Support provided within 8 business hours |
|  |  |  |  |

## Firebase by google

| Pros | Cons | Cost (looking from the cloud side of firebase) |
| --- | --- | --- |
| Provides all the database capabilities | Provide limited querying capabilities and as a result remain behind compare to other DB’s | Provides 100Gb of cloud storage & and 200 Gb of hi-res photos transferred at 23$/month (1800rs) |
| Provides a wide range of services and features | Provides limited data migration | If we want realtime database of 20GB then we have to pay 95$ extra (23+95= 118$) (9000rs (around)) |
| The Basic plan is Free | Is platform dependent |  |
| Quick and easy to integrate and setup  | Mainly android centered |  |
| Documentation is Concise and and provides every thing | And provides less support for iOS |  |
|  |  |  |

## Do Tell me if there are any mistakes in here

**Reply by atapas** [reply]
Thanks a lot, man, much appreciate it. Let me take a deep look at it and come back.

**Reply by hasnainmakada-99** [reply]
ok sir

**Comment by atapas** [comment]
I am inclined to select one between HarperDB and Firebase. More updates to follow.

**Reply by hasnainmakada-99** [reply]
Then sir you should go with harper db, as it is better than firebase

**Reply by atapas** [reply]
Tonight I'll be talking to people from HarperDB to understand the support part for our project. Let's see how that goes 😀

**Comment by atapas** [comment]
We are concluding the HarperDB as the database to go ahead with.

Advantages:

- Cloud-based instance
- Both SQL and NoSQL Supports
- Custom Functions
- Past experiences of using it.
- We will get some advantages with the plans, quota, and bandwidth(as per my discussion).

Thank You every one participating in the discussion of concluding it. I shall create relevant issues to,

- Configure the database instance.
- Get the existing play meta data into the DB
- Create a workflow to update the DB for new play meta data as part of the build process.
